http/events.go

```diff
@@ -0,0 +1,131 @@
+package http
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/hashicorp/vault/helper/namespace"
+	"github.com/hashicorp/vault/sdk/logical"
+	"github.com/hashicorp/vault/vault"
+	"google.golang.org/protobuf/encoding/protojson"
+	"nhooyr.io/websocket"
+)
+
+var eventTypeRegex = regexp.MustCompile(`.*/events/subscribe/(.*)`)
+
+// handleEventsSubscribeWebsocket runs forever, returning a websocket error code and reason
+// only if the connection closes or there was an error.
+func handleEventsSubscribeWebsocket(ctx context.Context, core *vault.Core, ns *namespace.Namespace, eventType logical.EventType, conn *websocket.Conn, json bool) (websocket.StatusCode, string, error) {
+	events := core.Events()
+	ch, err := events.Subscribe(ctx, ns, eventType)
+	if err != nil {
+		core.Logger().Info("Error subscribing", "error", err)
+		return websocket.StatusUnsupportedData, "Error subscribing", nil
+	}
+
+	defer close(ch)
+
+	for {
+		select {
+		case <-ctx.Done():
+			core.Logger().Info("Websocket context is done, closing the connection")
+			return websocket.StatusNormalClosure, "", nil
+		case message := <-ch:
+			core.Logger().Info("Got websocket message", "message", message)
+			var messageBytes []byte
+			if json {
+				messageBytes, err = protojson.Marshal(message)
+			} else {
+				messageBytes, err = proto.Marshal(message)
+			}
+			if err != nil {
+				core.Logger().Warn("Could not serialize websocket event", "error", err)
+				return 0, "", err
+			}
+			messageString := string(messageBytes) + "\n"
+			err = conn.Write(ctx, websocket.MessageText, []byte(messageString))
+			if err != nil {
+				return 0, "", err
+			}
+		}
+	}
+}
+
+func handleEventsSubscribe(core *vault.Core) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		core.Logger().Info("Got request to", "url", r.URL, "version", r.Proto)
+		matches := eventTypeRegex.FindStringSubmatch(r.URL.Path)
```

Comment:
Given we have a fixed length prefix, can we just do string indexing instead?

Reply:
Hmm. For some reason I thought that Vault could have an arbitrary prefix for requests, like, `/myprefix/v1/sys/events`, but maybe not?

Reply:
Oh you're right actually, we can have `/v1/<namespace>/sys/events` if we explicitly mount this API inside namespaces too (which we should). In ent, `genericWrappedHandler` ensures we should already have the namespace available in the context by the time we get to this handler so we could calculate the correct offset, maybe taking some extra care with the root namespace.

Although it's a bit of a hassle, I think it's worth it for resilience because we are not exerting any control over event type names, so it's pretty vulnerable to adversarial inputs as-is.